File: python/1-9/problem2.py
def even_fibonacci_numbers(n):
    '''Returns a list of even fibonacci numbers under 'n'.

    Args:
        n: The max value of a fibonacci number to be included.

    Returns:
        A list of all the even fibonacci numbers under 'n'
    '''
    fib = [1,1]
    while True:
        new_val = fib[-1] + fib[-2]
        if new_val > n:
            break
        fib.append(new_val)
    return [x for x in fib if x % 2 == 0]


if __name__ == '__main__':
    from timeit import timeit
    print(timeit(setup="from __main__ import even_fibonacci_numbers", stmt="sum(even_fibonacci_numbers(4000000))", number=1))

File: js/1-9/problem2.js
const prototypes = require('../prototypes');

// Returns an array of all Fibonacci numbers less than 'n'.
let fibonacciMax = function() {
    let fibonacci = [0, 1];

    return n => {
        let answer;
        for (let i = 1; i < Infinity; i++) {
            if (!fibonacci[i + 1]) {
                fibonacci.push(fibonacci[i] + fibonacci[i - 1]);
            }

            if (fibonacci[i + 1] > n) {
                answer = fibonacci.slice(0, i + 1);
                return answer;
            }
        }
    };
}();

console.log(fibonacciMax(4000000).evens().sum());
